src/datatype/svecf32.rs

```diff
@@ -0,0 +1,490 @@
+        if let Some(x) = option {
+            x
+        } else {
+            SessionError::BadLiteral {
+                hint: hint.to_string(),
+            }
+            .friendly()
+        }
+    }
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    enum State {
+        MatchingLeft,
+        Reading,
+        MatchedRight,
+    }
+    use State::*;
+    let input = input.to_bytes();
+    let mut vector = Vec::<SparseF32Element>::new();
+    let mut state = MatchingLeft;
+    let mut token: Option<String> = None;
+    let mut index = 0;
+    for &c in input {
+        match (state, c) {
+            (MatchingLeft, b'[') => {
+                state = Reading;
+            }
+            (Reading, b'0'..=b'9' | b'a'..=b'z' | b'A'..=b'Z' | b'.' | b'+' | b'-') => {
+                let token = token.get_or_insert(String::new());
+                token.push(char::from_u32(c as u32).unwrap());
+            }
+            (Reading, b',') => {
+                let token = solve(token.take(), "Expect a number.");
+                let value: F32 = solve(token.parse().ok(), "Bad number.");
+                if !value.is_zero() {
+                    vector.push(SparseF32Element { index, value });
+                }
+                index += 1;
+            }
+            (Reading, b']') => {
+                if let Some(token) = token.take() {
+                    let value: F32 = solve(token.parse().ok(), "Bad number.");
+                    if !value.is_zero() {
+                        vector.push(SparseF32Element { index, value });
+                    }
+                    index += 1;
+                }
+                state = MatchedRight;
+            }
+            (_, b' ') => {}
+            _ => {
+                SessionError::BadLiteral {
+                    hint: format!("Bad character with ascii {:#x}.", c),
+                }
+                .friendly();
+            }
+        }
+    }
+    if state != MatchedRight {
+        SessionError::BadLiteral {
+            hint: "Bad sequence.".to_string(),
+        }
+        .friendly();
+    }
+    if index > 65535 {
+        SessionError::BadValueDimensions.friendly();
+    }
+    SVecf32::new_in_postgres(&vector, index as u16)
+}
+
+#[pgrx::pg_extern(immutable, parallel_safe, strict)]
+fn _vectors_svecf32_out(vector: SVecf32Input<'_>) -> CString {
+    let mut buffer = String::new();
+    buffer.push('[');
+    let mut index = 0;
+    let mut iter = expand_sparse(&vector);
+    if let Some(x) = iter.next() {
+        buffer.push_str(format!("{}", x).as_str());
+        index += 1;
+    }
+    for x in iter {
+        buffer.push_str(format!(", {}", x).as_str());
+        index += 1;
+    }
+    while index < vector.dims() {
+        buffer.push_str(", 0");
+        index += 1;
+    }
+    buffer.push(']');
+    CString::new(buffer).unwrap()
+}
+
+#[pgrx::pg_extern(immutable, parallel_safe, strict)]
+fn svector_from_kv_string(dims: i32, input: &str) -> SVecf32Output {
```

Comment:
Do not expose public functions by `pg_extern`. Write schema manually.

Reply:
completed in new commit